File: website-analyzer.js
import axios from 'axios';
import * as cheerio from 'cheerio';
import fs from 'fs';
import path from 'path';

class WebsiteAnalyzer {
  constructor(baseUrl) {
    this.baseUrl = baseUrl;
    this.visited = new Set();
    this.brokenLinks = [];
    this.workingLinks = [];
    this.missingPages = [];
    this.externalLinks = [];
    this.images = [];
    this.scripts = [];
    this.stylesheets = [];
    this.maxDepth = 5;
    this.currentDepth = 0;
    this.sitemapUrls = [];
  }

  async analyze() {
    console.log(`🔍 Starting comprehensive analysis of ${this.baseUrl}`);
    console.log('=' .repeat(60));
    
    try {
      await this.loadSitemap();
      if (this.sitemapUrls.length > 0) {
        console.log(`🗺️  Found ${this.sitemapUrls.length} URLs in sitemap.xml`);
        for (const url of this.sitemapUrls) {
          await this.crawlPage(url, 0);
        }
      } else {
        console.log('⚠️  No sitemap URLs found; falling back to base URL crawl');
        await this.crawlPage(this.baseUrl, 0);
      }
      this.generateReport();
    } catch (error) {
      console.error('❌ Analysis failed:', error.message);
    }
  }

  async loadSitemap() {
    const sitemapUrl = new URL('/sitemap.xml', this.baseUrl).href;
    try {
      const response = await axios.get(sitemapUrl, {
        timeout: 10000,
        headers: { 'User-Agent': 'Mozilla/5.0 (compatible; ZionTechGroup-Bot/1.0)' }
      });
      if (response.status !== 200 || typeof response.data !== 'string') {
        return;
      }
      const locMatches = response.data.match(/<loc>(.*?)<\/loc>/g) || [];
      const urls = locMatches
        .map(tag => tag.replace('<loc>', '').replace('</loc>', '').trim())
        .filter(u => u.startsWith('http'));
      // Deduplicate and only include internal URLs
      const base = this.baseUrl.replace(/\/$/, '');
      this.sitemapUrls = Array.from(new Set(urls)).filter(u => u.startsWith(base));
    } catch (err) {
      console.warn(`⚠️  Failed to load sitemap.xml: ${err.message}`);
      this.sitemapUrls = [];
    }
  }

  async crawlPage(url, depth) {
    if (depth > this.maxDepth || this.visited.has(url)) {
      return;
    }

    this.visited.add(url);
    console.log(`📄 Crawling: ${url} (depth: ${depth})`);

    try {
      const response = await axios.get(url, {
        timeout: 10000,
        headers: {
          'User-Agent': 'Mozilla/5.0 (compatible; ZionTechGroup-Bot/1.0)'
        }
      });

      if (response.status === 200) {
        this.workingLinks.push(url);
        const $ = cheerio.load(response.data);
        
        // Extract all links
        const links = $('a[href]');
        links.each((i, element) => {
          const href = $(element).attr('href');
          if (href) {
            this.processLink(href, url, depth + 1);
          }
        });

        // Extract images
        const images = $('img[src]');
        images.each((i, element) => {
          const src = $(element).attr('src');
          if (src) {
            this.images.push({
              url: this.resolveUrl(src, url),
              page: url
            });
          }
        });

        // Extract scripts
        const scripts = $('script[src]');
        scripts.each((i, element) => {
          const src = $(element).attr('src');
          if (src) {
            this.scripts.push({
              url: this.resolveUrl(src, url),
              page: url
            });
          }
        });

        // Extract stylesheets
        const stylesheets = $('link[rel="stylesheet"], link[href*=".css"]');
        stylesheets.each((i, element) => {
          const href = $(element).attr('href');
          if (href) {
            this.stylesheets.push({
              url: this.resolveUrl(href, url),
              page: url
            });
          }
        });

        // Check for common missing content indicators
        this.checkMissingContent($, url);

      } else {
        this.brokenLinks.push({
          url,
          status: response.status,
          error: `HTTP ${response.status}`
        });
      }
    } catch (error) {
      this.brokenLinks.push({
        url,
        status: 'ERROR',
        error: error.message
      });
    }
  }

  processLink(href, baseUrl, depth) {
    const fullUrl = this.resolveUrl(href, baseUrl);
    
    if (fullUrl.startsWith(this.baseUrl)) {
      // Internal link
      if (!this.visited.has(fullUrl)) {
        setTimeout(() => this.crawlPage(fullUrl, depth), 100);
      }
    } else if (fullUrl.startsWith('http')) {
      // External link
      this.externalLinks.push({
        url: fullUrl,
        source: baseUrl
      });
    }
  }

  resolveUrl(href, baseUrl) {
    if (href.startsWith('http')) {
      return href;
    }
    
    try {
      return new URL(href, baseUrl).href;
    } catch {
      return href;
    }
  }

  checkMissingContent($, url) {
    const content = $.text();
    
    // Check for common missing content indicators
    if (content.includes('404') || content.includes('Page Not Found')) {
      this.missingPages.push({
        url,
        type: '404 Error'
      });
    }
    
    if (content.includes('Under Construction') || content.includes('Coming Soon')) {
      this.missingPages.push({
        url,
        type: 'Under Construction'
      });
    }
    
    if (content.length < 100) {
      this.missingPages.push({
        url,
        type: 'Minimal Content'
      });
    }
  }

  async checkExternalLinks() {
    console.log('\n🔗 Checking external links...');
    
    for (const link of this.externalLinks.slice(0, 20)) { // Limit to 20 for performance
      try {
        const response = await axios.head(link.url, { timeout: 5000 });
        if (response.status >= 400) {
          this.brokenLinks.push({
            url: link.url,
            status: response.status,
            error: `External link broken`,
            source: link.source
          });
        }
      } catch (error) {
        this.brokenLinks.push({
          url: link.url,
          status: 'ERROR',
          error: `External link error: ${error.message}`,
          source: link.source
        });
      }
    }
  }

  generateReport() {
    console.log('\n📊 Generating comprehensive analysis report...');
    
    const report = {
      timestamp: new Date().toISOString(),
      baseUrl: this.baseUrl,
      summary: {
        totalPages: this.workingLinks.length,
        brokenLinks: this.brokenLinks.length,
        missingPages: this.missingPages.length,
        externalLinks: this.externalLinks.length,
        images: this.images.length,
        scripts: this.scripts.length,
        stylesheets: this.stylesheets.length
      },
      workingLinks: this.workingLinks,
      brokenLinks: this.brokenLinks,
      missingPages: this.missingPages,
      externalLinks: this.externalLinks,
      images: this.images,
      scripts: this.scripts,
      stylesheets: this.stylesheets
    };

    // Save detailed report
    fs.writeFileSync('zion-website-analysis.json', JSON.stringify(report, null, 2));
    
    // Generate markdown report
    const markdownReport = this.generateMarkdownReport(report);
    fs.writeFileSync('zion-website-analysis-report.md', markdownReport);
    
    console.log('\n✅ Analysis complete! Reports generated:');
    console.log('   - zion-website-analysis.json (detailed data)');
    console.log('   - zion-website-analysis-report.md (readable report)');
    
    this.printSummary(report);
  }

  generateMarkdownReport(report) {
    return `# Zion Tech Group Website Analysis Report

**Generated:** ${new Date().toLocaleString()}
**Base URL:** ${report.baseUrl}

## 📊 Executive Summary

- **Total Pages Analyzed:** ${report.summary.totalPages}
- **Broken Links Found:** ${report.summary.brokenLinks}
- **Missing/Incomplete Pages:** ${report.summary.missingPages}
- **External Links:** ${report.summary.externalLinks}
- **Images:** ${report.summary.images}
- **Scripts:** ${report.summary.scripts}
- **Stylesheets:** ${report.summary.stylesheets}

## 🚨 Critical Issues

### Broken Links (${report.brokenLinks.length})
${report.brokenLinks.map(link => `- **${link.url}** - ${link.error}${link.source ? ` (from ${link.source})` : ''}`).join('\n')}

### Missing/Incomplete Pages (${report.missingPages.length})
${report.missingPages.map(page => `- **${page.url}** - ${page.type}`).join('\n')}

## ✅ Working Pages (${report.workingLinks.length})
${report.workingLinks.map(url => `- ${url}`).join('\n')}

## 🔗 External Links (${report.externalLinks.length})
${report.externalLinks.map(link => `- ${link.url} (from ${link.source})`).join('\n')}

## 🖼️ Images (${report.images.length})
${report.images.map(img => `- ${img.url}`).join('\n')}

## 📜 Scripts (${report.scripts.length})
${report.scripts.map(script => `- ${script.url}`).join('\n')}

## 🎨 Stylesheets (${report.stylesheets.length})
${report.stylesheets.map(css => `- ${css.url}`).join('\n')}

## 🎯 Recommendations

1. **Fix all broken links immediately**
2. **Complete missing page content**
3. **Verify external link validity**
4. **Optimize image loading**
5. **Review and optimize scripts and stylesheets**

---
*Report generated by Zion Tech Group Website Analyzer*
`;
  }

  printSummary(report) {
    console.log('\n📋 QUICK SUMMARY:');
    console.log('=' .repeat(40));
    console.log(`✅ Working Pages: ${report.summary.totalPages}`);
    console.log(`❌ Broken Links: ${report.summary.brokenLinks}`);
    console.log(`⚠️  Missing Pages: ${report.summary.missingPages}`);
    console.log(`🔗 External Links: ${report.summary.externalLinks}`);
    console.log(`🖼️  Images: ${report.summary.images}`);
    console.log(`📜 Scripts: ${report.summary.scripts}`);
    console.log(`🎨 Stylesheets: ${report.summary.stylesheets}`);
    
    if (report.summary.brokenLinks > 0) {
      console.log('\n🚨 CRITICAL: Broken links detected!');
      console.log('Top broken links:');
      report.brokenLinks.slice(0, 5).forEach(link => {
        console.log(`   - ${link.url}: ${link.error}`);
      });
    }
    
    if (report.summary.missingPages > 0) {
      console.log('\n⚠️  WARNING: Missing/incomplete pages detected!');
      console.log('Pages needing attention:');
      report.missingPages.slice(0, 5).forEach(page => {
        console.log(`   - ${page.url}: ${page.type}`);
      });
    }
  }
}

// Run the analysis
async function main() {
  const analyzer = new WebsiteAnalyzer('https://ziontechgroup.com');
  await analyzer.analyze();
}

// Check if this is the main module
const isMainModule = process.argv[1] === new URL(import.meta.url).pathname;
if (isMainModule) {
  main().catch(console.error);
}

export default WebsiteAnalyzer;

File: website_analyzer.py
#!/usr/bin/env python3
"""
Comprehensive website analyzer to check all links and identify broken ones
"""
import requests
import time
import re
from urllib.parse import urljoin, urlparse
from bs4 import BeautifulSoup
import json
from collections import defaultdict

class WebsiteAnalyzer:
    def __init__(self, base_url="https://ziontechgroup.com"):
        self.base_url = base_url
        self.visited_urls = set()
        self.broken_links = []
        self.working_links = []
        self.missing_pages = []
        self.all_links = set()
        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
        })
        
    def is_valid_url(self, url):
        """Check if URL is valid and belongs to the same domain"""
        try:
            parsed = urlparse(url)
            return parsed.netloc == urlparse(self.base_url).netloc or parsed.netloc == ''
        except:
            return False
    
    def normalize_url(self, url):
        """Normalize URL to handle relative paths"""
        if url.startswith('/'):
            return urljoin(self.base_url, url)
        elif not url.startswith('http'):
            return urljoin(self.base_url, '/' + url)
        return url
    
    def check_url(self, url):
        """Check if a single URL is accessible"""
        try:
            response = self.session.get(url, timeout=10, allow_redirects=True)
            if response.status_code == 200:
                return True, response.status_code
            else:
                return False, response.status_code
        except requests.exceptions.RequestException as e:
            return False, str(e)
    
    def extract_links(self, html_content, current_url):
        """Extract all links from HTML content"""
        soup = BeautifulSoup(html_content, 'html.parser')
        links = []
        
        # Extract all href attributes
        for link in soup.find_all('a', href=True):
            href = link['href']
            if self.is_valid_url(href):
                normalized_url = self.normalize_url(href)
                links.append({
                    'url': normalized_url,
                    'text': link.get_text(strip=True),
                    'source': current_url
                })
        
        # Extract all src attributes that might be links
        for img in soup.find_all('img', src=True):
            src = img['src']
            if self.is_valid_url(src):
                normalized_url = self.normalize_url(src)
                links.append({
                    'url': normalized_url,
                    'text': 'Image',
                    'source': current_url,
                    'type': 'image'
                })
        
        return links
    
    def analyze_page(self, url, depth=0, max_depth=3):
        """Analyze a single page and its links"""
        if url in self.visited_urls or depth > max_depth:
            return
        
        print(f"Analyzing: {url} (depth: {depth})")
        self.visited_urls.add(url)
        
        try:
            response = self.session.get(url, timeout=10, allow_redirects=True)
            if response.status_code == 200:
                self.working_links.append(url)
                self.all_links.add(url)
                
                # Extract links from this page
                links = self.extract_links(response.text, url)
                
                for link_info in links:
                    link_url = link_info['url']
                    self.all_links.add(link_url)
                    
                    # Check if it's a page link (not image)
                    if link_info.get('type') != 'image':
                        # Recursively analyze if it's a new page
                        if link_url not in self.visited_urls and depth < max_depth:
                            time.sleep(0.5)  # Be respectful to the server
                            self.analyze_page(link_url, depth + 1, max_depth)
            else:
                self.broken_links.append({
                    'url': url,
                    'status': response.status_code,
                    'error': f"HTTP {response.status_code}"
                })
        except requests.exceptions.RequestException as e:
            self.broken_links.append({
                'url': url,
                'status': 'ERROR',
                'error': str(e)
            })
    
    def find_missing_pages(self):
        """Find pages that are referenced but don't exist"""
        # This would need to be customized based on the site structure
        # For now, we'll check common page patterns
        common_pages = [
            '/about', '/contact', '/services', '/pricing', '/blog',
            '/privacy', '/terms', '/cookies', '/sitemap', '/team',
            '/careers', '/case-studies', '/demo', '/support'
        ]
        
        for page in common_pages:
            url = urljoin(self.base_url, page)
            if url not in self.all_links:
                is_working, status = self.check_url(url)
                if not is_working:
                    self.missing_pages.append({
                        'url': url,
                        'status': status,
                        'reason': 'Referenced but not accessible'
                    })
    
    def generate_report(self):
        """Generate a comprehensive analysis report"""
        report = {
            'summary': {
                'total_links_found': len(self.all_links),
                'working_links': len(self.working_links),
                'broken_links': len(self.broken_links),
                'missing_pages': len(self.missing_pages),
                'pages_analyzed': len(self.visited_urls)
            },
            'broken_links': self.broken_links,
            'missing_pages': self.missing_pages,
            'working_links': self.working_links[:50],  # Limit to first 50
            'all_links': list(self.all_links)
        }
        
        return report
    
    def run_analysis(self):
        """Run the complete website analysis"""
        print("Starting website analysis...")
        print(f"Base URL: {self.base_url}")
        
        # Start with the homepage
        self.analyze_page(self.base_url)
        
        # Find missing pages
        self.find_missing_pages()
        
        # Generate report
        report = self.generate_report()
        
        # Save report to file
        with open('website_analysis_report.json', 'w') as f:
            json.dump(report, f, indent=2)
        
        print(f"\nAnalysis complete!")
        print(f"Total links found: {report['summary']['total_links_found']}")
        print(f"Working links: {report['summary']['working_links']}")
        print(f"Broken links: {report['summary']['broken_links']}")
        print(f"Missing pages: {report['summary']['missing_pages']}")
        print(f"Pages analyzed: {report['summary']['pages_analyzed']}")
        
        return report

if __name__ == "__main__":
    analyzer = WebsiteAnalyzer()
    report = analyzer.run_analysis()
    
    # Print broken links
    if report['broken_links']:
        print("\n=== BROKEN LINKS ===")
        for link in report['broken_links']:
            print(f"❌ {link['url']} - {link['error']}")
    
    # Print missing pages
    if report['missing_pages']:
        print("\n=== MISSING PAGES ===")
        for page in report['missing_pages']:
            print(f"⚠️  {page['url']} - {page['reason']}")